### PR TITLE
Add support for SGR 58/59 (underline coloring)

### PR DESCRIPTION
**Pull request by tusharsnx**
## Summary of the Pull Request
Adds support for underline coloring with SGR 58/59 sequences. ConPTY can now parse, store and send those sequences across console applications and client terminal app. 

Tested it on Wezterm which already has support for drawing colored underlines. For testing, I replaced the prebundled `OpenConsole.exe` and `conpty.dll` from Wezterm with the modified versions.

```sh
printf "\n\e[4m\e[58;2;0;255;215mRGBCyanUnderline\e[24m \e[58;5;50m\e[4mINDEX256CyanUnderline\e[24m \e[4m\e[58;5;50m\e[38;2;255;0;95mUnderlineColorWithColoredText\e[24m \e[4m\e[38;5;50m\e[58;2;255;0;95mAnotherExample\n"
```

Above code produces output like so:

Left: Wezterm, Right: WindowsTerminal Dev
Screenshot 2023-06-25 183224

*Changes aren't visible on WindowsTerminal as of now. Atleast one of the graphic render, E.g. AtlasEngine, needs to support coloring of underline and drawing them on screen. Will be working on this in next few days.*

## PR Checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
- [ ] Schema updated (if necessary)

### CODE REVIEW

**Comment by tusharsnn** [comment]
@microsoft-github-policy-service agree

**Comment by tusharsnn** [comment]
Interestingly, running the example command on WT dev makes few words to blink (??), don't know what might cause this.

https://github.com/microsoft/terminal/assets/55626797/259094ab-ecb6-43c5-bafb-862c69847c7f

**Comment by tusharsnn** [comment]
Almost! 🤩

image

**Comment by tusharsnn** [comment]
>  in case an emulator doesn't recognize a sequence (let's say doesn't recognize the new extension of 58), subsequent numbers are interpreted as other attributes. E.g. if 256-color mode is chosen then the next numeric parameter is 5 which turns on blinking.

I guess, this [comment](https://github.com/neovim/neovim/issues/9270#issuecomment-441979176) explains the blinking issue, and why we need to prefer using `:` as the parameter sep for underline coloring.

File: src/renderer/vt/Xterm256Engine.cpp [hunk 1]
```diff
@@ -145,9 +146,77 @@ Xterm256Engine::Xterm256Engine(_In_ wil::unique_hfile hPipe,
...
+// Arguments:
+// - color: The color to emit a VT sequence for.
+// Return Value:
+// - S_OK if we succeeded, else an appropriate HRESULT for failing to allocate or write.
+[[nodiscard]] HRESULT Xterm256Engine::_UnderlineRGBColor(const COLORREF color) noexcept
+{
+    const auto r = GetRValue(color);
+    const auto g = GetGValue(color);
+    const auto b = GetBValue(color);
+    return _WriteFormatted(FMT_COMPILE("\x1b[58;2;{};{};{}m"), r, g, b);
```
[the remark this answers is not in the document]
Comment (tusharsnn) [reply src/renderer/vt/Xterm256Engine.cpp:212]: I see what you mean, may be we can do only `:` instead, but that would require removing the invalidation of `:` as a CSI parameter delimiter (conditionally). Thanks for the heads up.

**Comment by tusharsnn** [comment]
> Interestingly, running the example command on WT dev makes few words to blink (??), don't know what might cause this.

Ok, so there is no blinking issue in WT dev. I had changed the project location to another drive and forgot to re-deploy it from that location, so it was using the outdated one.

image

**Comment by tusharsnx** [comment]
closing this in favour of: https://github.com/microsoft/terminal/pull/15795

**Pull request closed without merging**